botocore/serialize.py

```diff
@@ -596,9 +597,64 @@ def _convert_header_value(self, shape, value):
         else:
             return value
 
+    def _finalize_serialization(self, serialized, operation_model):
+        """Perform any required post-processing steps for serialization."""
+        self._prepare_additional_traits(serialized, operation_model)
+
 
 class RestJSONSerializer(BaseRestSerializer, JSONSerializer):
 
+    def _finalize_serialization(self, serialized, operation_model):
+        super()._finalize_serialization(serialized, operation_model)
+
+        self._set_default_json_body(serialized, operation_model)
+        self._set_default_content_type_header(serialized, operation_model)
+        self._set_content_length(serialized, operation_model)
+
+    def _set_default_json_body(self, serialized, operation_model):
+        """Set default empty body ({}) for RestJSON requests if needed."""
+
+        # If we've already computed a body or will never serialize one,
+        # we can skip this. We will also skip modifying the body for
+        # blob or string payload shapes.
+        if (
+            serialized['body'] != b''
+            or not operation_model.can_serialize_body
+            or self._has_blob_or_string_payload(operation_model)
+        ):
+            return
+
+        serialized['body'] = b'{}'
+
+    def _set_default_content_type_header(self, serialized, operation_model):
+        """Set Content-Type to application/json for all structured bodies."""
+
+        # Don't set header if we have a payload
+        # that's either a blob or string
+        if self._has_blob_or_string_payload(operation_model):
+            return
+
+        has_body = serialized['body'] != b''
+        has_content_type = has_header('Content-Type', serialized['headers'])
+        if has_body and not has_content_type:
+            serialized['headers']['Content-Type'] = 'application/json'
+
+    def _set_content_length(self, serialized, operation_model):
+        headers = serialized['headers']
+        has_content_length = has_header('Content-Length', headers)
+
+        if not has_content_length and operation_model.can_serialize_body:
```

Comment:
Why not just check if we have a non-zero length body? At this point we've actually ran the serializer and the serializer has produced a body, whether or not the model conceptually says a body is possible doesn't seem relevant anymore.

Reply:
We don't want to emit a content-length when it's not relevant, e.g. if it's a GET request we shouldn't be emitting the Content-Type. This matches our current behavior in httpsession.py which would deviate without this check.